test/drivers/test_cqrrp.cc

```diff
@@ -0,0 +1,228 @@
+        T col_norm = 0.0;
+        int max_idx = 0;
+        for(int i = 0; i < n; ++i) {
+            col_norm = blas::nrm2(m, &A_dat[m * i], 1);
+            if(max_col_norm < col_norm) {
+                max_col_norm = col_norm;
+                max_idx = i;
+            }
+        }
+        T col_norm_A = blas::nrm2(n, &A_cpy_dat[m * max_idx], 1);
+        T norm_AQR = lapack::lange(Norm::Fro, m, n, A_dat, m);
+        
+        printf("REL NORM OF AP - QR:    %14e\n", norm_AQR / norm_A);
+        printf("MAX COL NORM METRIC:    %14e\n", max_col_norm / col_norm_A);
+        printf("FRO NORM OF (Q'Q - I):  %14e\n\n", norm_0 / std::sqrt((T) n));
+
+        T atol = std::pow(std::numeric_limits<T>::epsilon(), 0.75);
+        ASSERT_NEAR(norm_AQR / norm_A,         0.0, atol);
+        ASSERT_NEAR(max_col_norm / col_norm_A, 0.0, atol);
+        ASSERT_NEAR(norm_0, 0.0, atol);
+    }
+
+    /// General test for CQRRPT:
+    /// Computes QR factorzation, and computes A[:, J] - QR.
+    template <typename T, typename RNG, typename alg_type>
+    static void test_CQRRP_general(
+        T d_factor, 
+        T norm_A,
+        CQRRPTestData<T> &all_data,
+        alg_type &CQRRP,
+        RandBLAS::RNGState<RNG> &state) {
+
+        auto m = all_data.row;
+        auto n = all_data.col;
+
+        CQRRP.call(m, n, all_data.A.data(), m, d_factor, all_data.tau.data(), all_data.J.data(), state);
+        all_data.rank = CQRRP.rank;
+        
+        RandLAPACK::util::upsize(all_data.rank * n, all_data.R);
+        lapack::lacpy(MatrixType::Upper, all_data.rank, n, all_data.A.data(), m, all_data.R.data(), all_data.rank);
+
+        
+        lapack::ungqr(m, n, n, all_data.A.data(), m, all_data.tau.data());
+        
+        lapack::lacpy(MatrixType::General, m, all_data.rank, all_data.A.data(), m, all_data.Q.data(), m);
+
+        printf("RANK AS RETURNED BY CQRRP %4ld\n", all_data.rank);
+
+        RandLAPACK::util::col_swap(m, n, n, all_data.A_cpy1.data(), m, all_data.J);
+        RandLAPACK::util::col_swap(m, n, n, all_data.A_cpy2.data(), m, all_data.J);
+
+        error_check(norm_A, all_data);
+    }
+};
+
+// Note: If Subprocess killed exception -> reload vscode
+TEST_F(TestCQRRP, CQRRP_blocked_full_rank_basic) {
+    int64_t m = 60;//5000;
+    int64_t n = 60;//2000;
+    int64_t k = 60;
+    double d_factor = 1;//1.0;
+    int64_t b_sz = 10;//500;
+    double norm_A = 0;
+    double tol = std::pow(std::numeric_limits<double>::epsilon(), 0.85);
+    auto state = RandBLAS::RNGState();
+
+    CQRRPTestData<double> all_data(m, n, k);
+    RandLAPACK::CQRRP_blocked<double, r123::Philox4x32> CQRRP_blocked(true, tol, b_sz);
+    CQRRP_blocked.nnz = 2;
+    CQRRP_blocked.num_threads = 4;
+
+    RandLAPACK::gen::mat_gen_info<double> m_info(m, n, RandLAPACK::gen::gaussian);
+    //RandLAPACK::gen::mat_gen_info<double> m_info(m, n, RandLAPACK::gen::polynomial);
+    //m_info.cond_num = 2;
+    //m_info.rank = k;
+    //m_info.exponent = 2.0;
```

Comment:
This commented-out code should be removed, unless you have a specific reason for keeping it.

Reply:
Yes, this is in case i decide to change test matrix generator and so that i don't forget what parameters should be used with polynomial decay.